src/value.jl

```diff
@@ -1,145 +1,174 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
-end
-
-Base.convert(::Type{CategoricalValue{T, R}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{CategoricalValue{T}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{CategoricalValue}, x::CategoricalValue) = x
-
-Base.convert(::Type{Union{CategoricalValue{T, R}, Null}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{Union{CategoricalValue{T}, Null}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
-
-Base.promote_rule(::Type{CategoricalValue{S, R}}, ::Type{T}) where {S, T, R} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue{S}}, ::Type{T}) where {S, T} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue}, ::Type{T}) where {T} = T
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# checks whether the type is "categorical value"
+iscatvalue(::Type) = false
+iscatvalue(::Type{Union{}}) = false # otherwise it dispatches to Type{<:...}
+iscatvalue(::Type{<:CatValue}) = true
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+leveltype(::Type{<:CategoricalValue{T}}) where {T} = T
+leveltype(::Type{<:CategoricalString}) = String
+leveltype(::Type) = throw(ArgumentError("Not a \"categorical value\" type"))
+leveltype(x::Any) = leveltype(typeof(x))
+# eltype() is a synonym of leveltype() for categorical values
+Base.eltype(::Type{T}) where {T <: CatValue} = leveltype(T)
+Base.eltype(x::CatValue) = eltype(typeof(x))
+
+# integer type of category reference codes used by categorical value
+reftype(::Type{<:CatValue{R}}) where {R} = R
+reftype(x::Any) = reftype(typeof(x))
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvaluetype(::Type{T}) where {T} = T
+unwrap_catvaluetype(::Type{T}) where {T >: Null} =
+    Union{unwrap_catvaluetype(Nulls.T(T)), Null}
+unwrap_catvaluetype(::Type{Null}) = Null # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{Any}) = Any # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{T}) where {T <: CatValue} = leveltype(T)
+
+# get the "categorical value" type given value type `T` and reference type `R`
+catvaluetype(::Type{T}, ::Type{R}) where {T >: Null, R} =
+    catvaluetype(Nulls.T(T), R)
+catvaluetype(::Type{T}, ::Type{R}) where {T <: CatValue, R} =
+    reftype(T) === R ? T : catvaluetype(leveltype(T), R)
```

Comment:
Repeating my comment from above: Still the same  remark about `reftype(T) === R`.

Reply:
This is correct now and this would be correct if categorical values are extended (without this check there could be unnecessary copies / unexpected changes of behaviour).
I suggest to keep it as is.

Reply:
But in the even categorical values were extended, we wouldn't want the return type to change depending on whether `R` is the same or not. Imagine `CategoricalString` was only provided by an external package: would you like to have `catvaluetype(CategoricalString{UInt8}, UInt8) == CategoricalString{UInt8}` but `catvaluetype(CategoricalString{UInt8}, UInt32) == CategoricalValue{String, UInt32}`? Custom categorical value types would have to override `catvaluetype` anyway. So better make it very clear, rather than having a different behavior only in a very specific case which is less likely to be spotted.

Reply:
> Custom categorical value types would have to override `catvaluetype()` anyway. 

Exactly (otherwise they won't be used), but not the `T<:CatValue` method.
So there should be no situation you describe.

Though the situation that I'm referring to (multiple categorical value implementations for the same level type that are selected depending on the context) are not supported by the current API.
So I'll drop `R` check to make everyone happy.